examples/nn.dx

```diff
@@ -0,0 +1,227 @@
+' # Neural Networks
+
+include "plot.dx"
+
+' ## NN Prelude
+
+
+def relu (input : Float) : Float =
+  select (input > 0.0) input 0.0
+
+instance [Add a, Add b] Add (a & b)
+  add = \(a, b) (c, d). ( (a + c), (b + d))
+  sub = \(a, b) (c, d). ( (a - c), (b - d))
+  zero = (zero, zero)
+
+instance [VSpace a, VSpace b] VSpace (a & b)
+  scaleVec = \ s (a, b) . (scaleVec s a, scaleVec s b)
+
+def Layer (inp:Type) (out:Type) (param:Type) : Type=
+  {forward:(param -> inp -> out) &
+   init: Key -> param
+  }
+
+def forward (l:Layer i o p) (p : p) (x : i): o =
+  (getAt #forward l) p x
+
+def init (l:Layer i o p) (k:Key)  : p  =
+  (getAt #init l) k
+
+
+' ## Layers 
+
+' Dense layer
+
+def Dense (a:Type) (b:Type) : Type =
+  Layer (a=>Float) (b=>Float) ((a=>b=>Float) & (b=>Float))
+
+def dense (a:Type) (b:Type) : Dense a b =
+  {
+   forward = (\(weight, bias) x .
+               for j:b. (bias.j + sum for i :a. weight.i.j * x.i)),
```

Comment:
I love how this doesn't use matmul.
I love that increase to clarity

I be it suffers from the same underlying problem that makes matmul slow, and that that is one of the main things driving the performance slowness.
#346

Reply:
> I love that increase to clarity

This is actually profound. Woah.

`matmul` isn't unclear either though, but it's subject to the pitfall of implicit broadcasting? `matmul(weight, x) + bias`

Reply:
I am not sure if matmul is unclear or not.
I suspect it actually is; but we have twisted our minds around it enough that we find it clear.

I am thinking back to me as a naive undergrad implementing a Neural net in Java,
where I had objects called Neurons, with a scalar field called `weight`.
which doesn't end up at good places because performance.

Reply:
You're right. Being more explicit and "pointy" (as opposed to pointfree or using abstractions) should be easier to understand, as long as the per-scalar innermost loop body isn't too complicated.

Reply:
Yeah, I kind of dig this too if it's fast. 

However writing relu as `for i. for j. for k. relu.i.j.k` is a bit annoying. I might type class that.

Reply:
`for i j k. relu x.i.j.k` is better and i think works?

Far more speculative I wonder it we can have some kind of nesting notion.
Maybe some kind multindeing type, and having repeated dots indicate how many time to unpeal that multiindexing type
so that `for i. relu x...i` is same as `for i j k. real x.i.j.k`
idk if that gives up the carity though.

Reply:
Lol, I had no idea I could do that, despite lambdas working that way. That is way better. 

Multiindexing is a pretty fun idea.

Reply:
Fixed.